Contributing guide doc (#170)
gauge-sh/tach: docs/contributing/setting-up-project.mdx

---
title: Setting up the Project for Contributors
---
Tach CLI is an open source 

#### 1. Fork the repository

#### 2. Clone the repository locally
```bash 
git clone https://github.com/github_username/tach.git
cd tach
```
It is adviced to work on a branch of your own and then raise a PR
```bash
git checkout -b "your_feature_name" 
```

#### 3. Create a virtual environment
```bash
python3 -m venv env
```
Activate the environment 
```bash
source env/Scripts/activate

> ```

Reviewer: These sections should be removed, since we should assume open-source contributors are able to fork the repo and open a pull request. There might be a standard GitHub documentation link we can provide, but we shouldn't need to explain it from scratch.

For the virtual environment, I'll add more detail in other comments but this is managed automatically by our `Makefile`

Author: even the dev-requirements is handled in Makefile, removing it from the doc

#### 4. Install Dependencies
```bash
pip install -r dev-requirements.txt
```

#### 5. Build

This build will reflect changes that are currently existing locally

To build the project - run:

```bash
 pip install .
```

> Note: tach is a CI tool to enforce dependencies, using modular architecture. For testing your changes you'll need to set this project up within your own project

Reviewer: We should recommend that contributors start by running: `make deps`

This will automatically create and activate a virtual environment using `uv`, as well as build and install the project.

To rebuild after changes to Rust files, we should recommend using `make install`

See here: https://github.com/gauge-sh/tach/blob/main/Makefile

#### 6. Test

Tach uses `pytest` module for testing all the files within `python/tests/`
```bash
pytest relative_path_of_file_or_folder_to_test/ # To test everything just run `pytest`

> ```

Reviewer: This can instead suggest running `make test`

#### 7. Setting up the docs
Tach uses mintlify platform to create and maintain public facing documentation
```bash 
npm install -g mintlify
cd docs
mintlify dev # Build docs for a local preview

> ```

Reviewer: This can suggest running  `make docs`, but you should mention that the contributor will need to install Node and NPM on their own.

### 8. Things to check before committing
Sync all your dependencies in the root folder
```bash
tach sync
```
Ruff checks (Automatic Linting)
```bash 
ruff check --fix
ruff format --check 
ruff format

> ```

Reviewer: We should suggest:

```
tach check
make type-check
make lint
make fmt
```

That's it! You are now ready to push your new dev branch to your forked repo and then raise a PR with appropriate description

Find Beginner Friendly issues here: 
- [Good First Issues (For beginners)](https://github.com/gauge-sh/tach/issues?q=is%3Aopen+is%3Aissue+label%3A%22good+first+issue%22)
- [Documentation Issues](https://github.com/gauge-sh/tach/issues?q=is%3Aopen+is%3Aissue+label%3Adocumentation)
- [Issues](https://github.com/gauge-sh/tach/issues)
- [Documentation](https://github.com/gauge-sh/tach/tree/main/docs)

> - For any queries, just drop a message on our discord channel : [Join our Discord Channel](https://discord.com/invite/a58vW8dnmw)

Reviewer: ```suggestion

For any questions, just drop a message in [Discord](https://discord.com/invite/a58vW8dnmw)
```
